feat: add session Jupyter free images - Renku 2.0

andre-code opened the pull request:
PR for Jupyter free images - Renku 2.0


- [x] Support custom environment variables in Renku 2.0 - [#3227](https://github.com/SwissDataScienceCenter/renku-ui/pull/3227)
- [x] Use the data-service sessions endpoint - [#3296](https://github.com/SwissDataScienceCenter/renku-ui/pull/3296)
- [x] Resolve issues with data connectors 
- [x] Display an error message for session launch failures
- [x] Show an error when the image doesn't exist during session launch 

/deploy renku=feat-jupyter-free-sessions renku-data-services=main amalthea-sessions=main renku-notebooks=master extra-values=amalthea-sessions.deployCrd=false

leafty reviewed (commented):
It is possible that the double POST bug is present in this PR:

1. POST with empty list of data connectors
2. POST with data connectors

leafty reviewed (commented):
![Screenshot 2024-10-30 at 14 40 30](https://github.com/user-attachments/assets/2e3ff6dd-edaa-421a-91a5-d0010a8178c5)

Rename "fragment" to "path" (or "pathname"), URL component have specific names we have to adhere to -> `URI = scheme ":" ["//" authority] path ["?" query] ["#" fragment]`. See also: https://developer.mozilla.org/en-US/docs/Web/API/URL.

Also, the UI is not respecting this setting. The `<iframe>` needs to contain the URL path setting, otherwise it cannot support JupyterLab. I can open the interface with "Open in new tab" and manually adding the path there.

andre-code commented:
Thank you @leafty, for reviewing the PR and reporting the errors:

Visual bug: Fixed
Double POST issue: Fixed
Error in showing the session or opening the URL link: I'm holding this for now to see if Tasko can include the fix in the data service; otherwise, I can retrieve the default URL from the launcher and make it work.

andre-code commented:
> > Error in showing the session or opening the URL link: I'm holding this for now to see if Tasko can include the fix in the data service; otherwise, I can retrieve the default URL from the launcher and make it work.
> 
> ~Why does this need a backend fix? It is purely a UI issue.~ Nevermind.

@leafty Tasko will add the defaultURL to the backend GET /sessions response

leafty commented on client/src/components/Logs.tsx:
```diff
@@ -353,38 +367,24 @@ const EnvironmentLogs = ({ name, annotations }: EnvironmentLogsProps) => {
  */
 interface EnvironmentLogsPresentProps {
-  annotations: Record<string, unknown>;
+  title: ReactNode | string;
```
```suggestion
  title: ReactNode;
```

`string` is already part of `ReactNode`.

andre-code replied:
@leafty  fixed in the latest commit

andre-code commented:
Note:
🐞 Custom environments are currently being created as GLOBAL. This issue is awaiting a fix in the backend.

leafty commented:
The UI seems OK, but there are still issues with starting sessions, e.g. 
![Screenshot_20241106-124105](https://github.com/user-attachments/assets/3f659278-2cc2-4ac5-b989-75c84d0356c5)

leafty reviewed (approved):
LGTM

Please, do not merge until the whole feature is ready.
